### PR TITLE
feat: Improved Fee Rate Computation

### PR DESCRIPTION
**Pull request by gregorycoppola**
Addresses #2954

https://github.com/blockstack/stacks-blockchain/issues/2954#issuecomment-983017453 discusses four proposals for improving fee estimation.

We address these first three:

1. Address outlier block sensitivity by using a median of a window, rather than exponential windowing.
2. Address impact of STX transfers by using percentiles weighted by the proportion of the block limit consumed rather than 5th, 50th, and 95th percentiles. See weighted percentiles.
3. Address impact of non-full blocks by treating the non-full proportion of the block as a transaction at the relay fee minimum (a fee of 0 is too low for admission to the mempool).

These changes are tested with a suite of new unit tests.

### CODE REVIEW

File: src/cost_estimates/fee_medians.rs [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        Self::instantiate_db(&tx)?;
+        tx.commit()?;
+
+        Ok(Self {
+            db,
+            metric,
+            window_size,
+            full_block_weight: 6 * PROPORTION_RESOLUTION,
+        })
+    }
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:91]: Can you also update the open behavior in `cost_estimates::fee_scalar::ScalarFeeRateEstimator` so that it behaves the same as this open (basically, on every open, invoke `instantiate_db` to check if the table exists). The reason to do this check rather than the one it currently does (`instantiate` only if `OPEN_CREATE` was necessary) is that the node configuration will use the same SQLite database for both of these estimators, but the estimators only create their own tables (so the database may exist, but not have the table for the opened estimator yet).

File: src/cost_estimates/fee_medians.rs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+/// 3) "Pad" the block, so that any unused spaces is considered to have an associated fee rate of
+///    1f, the minimum. Ignoring the amount of empty space leads to over-estimates because it
+///    ignores the fact that there was still space in the block.
+pub struct WeightedMedianFeeRateEstimator<M: CostMetric> {
+    db: Connection,
+    /// We only look back `window_size` fee rates when averaging past estimates.
+    window_size: u32,
+    /// The weight of a "full block" in abstract scalar cost units. This is the weight of
+    /// a block that is filled on each dimension.
+    full_block_weight: u64,
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:57]: I think that the "fill with empty" behavior is a little more honest if the "full" block weight is  a block with just _one_ dimension filled. Because a block cannot exceed any dimension, a block with a weight > a single dimension fill may not be able to accept any more transactions. Currently, I think this only impacts the docstring comment, but when it's integrated into the `testnet/stacks-node`, that will impact how this value is set.

File: src/cost_estimates/fee_medians.rs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+///
+/// `sorted_fee_rates` must be non-empty.
+fn fee_rate_estimate_from_sorted_weighted_fees(
+    sorted_fee_rates: &Vec<FeeRateAndWeight>,
+) -> FeeRateEstimate {
+    if sorted_fee_rates.is_empty() {
+        panic!("`sorted_fee_rates` cannot be empty.");
+    }
+
+    let mut total_weight = 0u64;
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:245]: You can avoid worrying about overflows in `total_weight` and `cumulative_weight` by using `f64` for them (especially because they get cast to f64 anyways)

File: src/cost_estimates/fee_medians.rs [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    }
+    let mut cumulative_weight = 0u64;
+    let mut percentiles = Vec::new();
+    for rate_and_weight in sorted_fee_rates {
+        cumulative_weight += rate_and_weight.weight;
+        let percentile_n: f64 =
+            (cumulative_weight as f64 - rate_and_weight.weight as f64 / 2f64) / total_weight as f64;
+        percentiles.push(percentile_n);
+    }
+
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:257]: Can you add an `assert_eq!` here for `percentiles.len()` and `sorted_fee_rates.len()`?

File: src/cost_estimates/fee_medians.rs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    let target_percentiles = vec![0.05, 0.5, 0.95];
+    let mut fees_index = 0; // index into `sorted_fee_rates`
+    let mut values_at_target_percentiles = Vec::new();
+    for target_percentile in target_percentiles {
+        while fees_index < percentiles.len() && percentiles[fees_index] < target_percentile {
+            fees_index += 1;
+        }
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:264]: What happens when `percentiles[fees_index] == target_percentile`?

I think in the correct operation of the weighted percentile algorithm, `value` should equal `sorted_fee_rates[fee_index].fee_rate` in that case. Is that the case in this implementation?

File: src/cost_estimates/fee_medians.rs [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        low: values_at_target_percentiles[0],
+    }
+}
+
+/// If the weights in `working_rates` do not add up to `full_block_weight`, add a new entry **in
+/// place** that takes up the remaining space.
+fn maybe_add_minimum_fee_rate(working_rates: &mut Vec<FeeRateAndWeight>, full_block_weight: u64) {
+    let mut total_weight = 0u64;
+    for rate_and_weight in working_rates.into_iter() {
+        total_weight += rate_and_weight.weight;
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:292]: You should avoid an overflow here by doing checked_add:

```rust
total_weight = match total_weight.checked_add(rate_and_weight.weight) {
   Some(result) => result
   None => return,
};
```

Which just returns from this function if there's an overflow (because if there's an overflow the block definitely does not need any filling help).

File: src/cost_estimates/fee_medians.rs [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+        //  which would not have created the necessary table for this estimator.
+        let tx = tx_begin_immediate_sqlite(&mut db)?;
+        Self::instantiate_db(&tx)?;
+        tx.commit()?;
+
+        Ok(Self {
+            db,
+            metric,
+            window_size,
+            full_block_weight: 6 * PROPORTION_RESOLUTION,
```
Comment (kantai) [inline src/cost_estimates/fee_medians.rs:89]: Regarding my comment above about the block weight to use -- I think this should just be a single dimension fill (so just `PROPORTION_RESOLUTION`), because a block with that weight may not be able to add any more transactions.

**Review by kantai — COMMENTED** [review]
This looks mostly good to me, I had a couple comments in the current implementation, and I think you should have a test case for some of the corner cases of weighted percentiles (i.e., the percentile requested is exactly equal to one of the percentiles, explicit test cases for capturing too-low of a target percentile and too-high of a target percentile).

Additionally, you should integrate this fee estimator into the testnet/stacks-node configuration: the `config.rs` file handles instantiating estimators, and this new estimator should be added there (and probably set to the default fee estimator). And please also add an entry to the `CHANGELOG.md`, under a new `## [Unreleased]` subsection.

File: src/cost_estimates/fee_medians.rs (hunk 5 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs:264]: I exposed `fee_rate_estimate_from_sorted_weighted_fees` for test, and added test cases in `src/cost_estimates/tests/fee_medians.rs`.

@kantai I just put the answers from the program in to the right-hand side of the "expect".. What do you think?

File: src/cost_estimates/fee_medians.rs (hunk 6 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs:292]: done

File: src/cost_estimates/fee_medians.rs (hunk 1 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs:91]: copied.. we'll see if it causes any regressions in tests.

File: src/cost_estimates/fee_medians.rs (hunk 3 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs:245]: done in latest

File: src/cost_estimates/fee_medians.rs (hunk 4 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs:258]: done in latest

File: src/cost_estimates/fee_medians.rs (hunk 7 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs]: dropped the 6. some of the parameters in the tests changed, because the blocks got different amounts full.

File: src/cost_estimates/fee_medians.rs (hunk 2 above)
Comment (gregorycoppola) [reply src/cost_estimates/fee_medians.rs]: changed the text to say its any dimension

**Comment by gregorycoppola** [comment]
see https://github.com/blockstack/stacks-blockchain/pull/2972
